File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P230_KthSmallestElementInABST.py
# author: tobebetter9527
# since: 2024/3/18 10:42
from typing import Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.k = k
        self.inOrder(root)
        return self.ans

    def inOrder(self, root):
        if not root:
            return
        self.inOrder(root.left)
        if self.k == 0:
            return
        self.k -= 1
        if self.k == 0:
            self.ans = root.val
        self.inOrder(root.right)

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong2/binary/tree/P230_KthSmallestElementInABST.java
package com.dsa.leetcode.labuladong2.binary.tree;

public class P230_KthSmallestElementInABST {
    int res;
    int rank;

    public int kthSmallest(TreeNode root, int k) {
        traversal(root, k);
        return res;
    }

    private void traversal(TreeNode root, int k) {
        if (root == null) {
            return;
        }
        traversal(root.left, k);
        rank++;
        if (rank == k) {
            res = root.val;
            return;
        }
        traversal(root.right, k);
    }
}
